adding cluster loader to snafu

aakarshg opened the pull request:
Adding cluster loader to snafu

aakarshg commented:
> Can you provide a detailed dashboard since you own the CL work? We need someone who understands the workload to build an end to end vs just a tool to warehouse data.
> 

I don't understand the workload in depth, I only have enough understanding to get this pr in. 

> This current dashboard yields a nested document, do you happen to have a Grafana dashboard using this?

No.

jtaleric commented on cluster_loader/trigger_cluster_loader.py:
```diff
@@ -0,0 +1,86 @@
+        output_template['cluster_name'] = self.cluster_name
+        output_template['uuid'] = self.uuid
+        output_template['user'] = self.user
+        output_template['sample'] = self.sample
```
We need to iterate through some testing to ensure this is the right structure to build visuals from.

jtaleric commented on cluster_loader/README.md:
````diff
@@ -0,0 +1,42 @@
+```
+
+Note: additional arguments that can be passed are:
+
````
Formatting issue with the below.

aakarshg replied:
seems to be [fixed](https://github.com/cloud-bulldozer/snafu/blob/656adc98e8068bdf7cc06874d3600b4f2096894c/cluster_loader/README.md)

jtaleric commented:
Ran against my AWS cloud: 

```
jtaleric at bolt in ~/code/snafu on (no branch)△ $ python2.7 run_snafu.py -t cl -s 1 -p /home/jtaleric/go/src/github.com/openshift/origin/_output/local/bin/linux/amd64/openshift-tests 
2019-11-12T14:40:00Z - INFO     - MainProcess - run_snafu: logging level is INFO
2019-11-12T14:40:00Z - INFO     - MainProcess - wrapper_factory: identified cl as the benchmark wrapper
2019-11-12T14:40:00Z - INFO     - MainProcess - trigger_cluster_loader: running:/home/jtaleric/go/src/github.com/openshift/origin/_output/local/bin/linux/amd64/openshift-tests run-test "[Feature:Performance][Serial][Slow] Load cluster should load the cluster [Suite:openshift]"
2019-11-12T14:40:01Z - INFO     - MainProcess - trigger_cluster_loader: from current directory /home/jtaleric/code/snafu
2019-11-12T14:51:34Z - INFO     - MainProcess - trigger_cluster_loader: completed sample 1 , results in /home/jtaleric/code/1/cl_output.txt
2019-11-12T14:51:34Z - INFO     - MainProcess - trigger_cluster_loader: cl output is {"marker":"cluster_loader_marker","name":"cluster-loader-test","type":"metrics.TestDuration","startTime":"2019-11-12T14:40:04.723674137-05:00","testDuration":688895368891,"Steps":[{"type":"metrics.TemplateStepDuration","startTime":"2019-11-12T14:40:05.348133831-05:00","totalTime":2672228316,"rateDelay":0,"rateDelayCount":0,"stepPause":0,"stepPauseCount":0,"syncTime":480},{"type":"metrics.TemplateStepDuration","startTime":"2019-11-12T14:40:08.62952041-05:00","totalTime":1599381465,"rateDelay":0,"rateDelayCount":0,"stepPause":0,"stepPauseCount":0,"syncTime":228},{"type":"metrics.TemplateStepDuration","startTime":"2019-11-12T14:40:10.831949859-05:00","totalTime":1642830572,"rateDelay":0,"rateDelayCount":0,"stepPause":0,"stepPauseCount":0,"syncTime":347},{"type":"metrics.TemplateStepDuration","startTime":"2019-11-12T14:40:13.082772848-05:00","totalTime":1652103253,"rateDelay":0,"rateDelayCount":0,"stepPause":0,"stepPauseCount":0,"syncTime":423},{"type":"metrics.TemplateStepDuration","startTime":"2019-11-12T14:40:15.34071092-05:00","totalTime":1589622893,"rateDelay":0,"rateDelayCount":0,"stepPause":0,"stepPauseCount":0,"syncTime":588}]}
Traceback (most recent call last):
  File "run_snafu.py", line 127, in <module>
    sys.exit(main())
  File "run_snafu.py", line 89, in main
    for i in process_generator(index_args, parser):
  File "run_snafu.py", line 109, in process_generator
    for action, index in data_object.emit_actions():
  File "/home/jtaleric/code/snafu/cluster_loader/trigger_cluster_loader.py", line 60, in emit_actions
    config_file_location = list(filter(pattern.match, output_file_content))[0].strip().split('INFO: Using config ',1)[1].replace('"','')
IndexError: list index out of range

```

aakarshg commented:
Can you share the configuration you've used ? as well as the file /home/jtaleric/code/1/cl_output.txt

jtaleric commented:
Hit another failure 
```
jtaleric at bolt in ~/code/snafu on (no branch)▲ $ python2.7 run_snafu.py -t cl -s 1 -p /home/jtaleric/go/src/github.com/openshift/origin/_output/local/bin/linux/amd64/openshift-tests 
2019-11-12T15:22:18Z - INFO     - MainProcess - run_snafu: logging level is INFO
2019-11-12T15:22:18Z - INFO     - MainProcess - wrapper_factory: identified cl as the benchmark wrapper
2019-11-12T15:22:18Z - INFO     - MainProcess - trigger_cluster_loader: running:/home/jtaleric/go/src/github.com/openshift/origin/_output/local/bin/linux/amd64/openshift-tests run-test "[Feature:Performance][Serial][Slow] Load cluster should load the cluster [Suite:openshift]"
2019-11-12T15:22:18Z - INFO     - MainProcess - trigger_cluster_loader: from current directory /home/jtaleric/code/snafu
fail [github.com/openshift/origin/test/extended/cluster/utils.go:192]: Nov 12 15:39:38.437: Error in pod wait... timed out waiting for the condition
2019-11-12T15:39:38Z - ERROR    - MainProcess - trigger_cluster_loader: Command ' /home/jtaleric/go/src/github.com/openshift/origin/_output/local/bin/linux/amd64/openshift-tests run-test "[Feature:Performance][Serial][Slow] Load cluster should load the cluster [Suite:openshift]"' returned non-zero exit status 1
Traceback (most recent call last):
  File "/home/jtaleric/code/snafu/cluster_loader/trigger_cluster_loader.py", line 41, in emit_actions
    process = subprocess.check_call(command_string, stdout=file_stdout, shell=True)
  File "/usr/lib64/python2.7/subprocess.py", line 190, in check_call
    raise CalledProcessError(retcode, cmd)
CalledProcessError: Command ' /home/jtaleric/go/src/github.com/openshift/origin/_output/local/bin/linux/amd64/openshift-tests run-test "[Feature:Performance][Serial][Slow] Load cluster should load the cluster [Suite:openshift]"' returned non-zero exit status 1

```

my viperconfig 
```
---
provider: local
ClusterLoader:
  cleanup: true
  projects:
    - num: 1
      basename: podvert
      tuning: default
      ifexists: delete
      pods:
        - num: 1000
          basename: podvert
          file: podvert-pod.yaml
  tuningsets:
    - name: default
      pods:
        stepping:
          stepsize: 500
          pause: 30
          timeout: 600
        ratelimit:
          delay: 0

```

podvert-pod.yaml
```
---
kind: Pod
apiVersion: v1
metadata:
  name: podvert
  labels:
    name: podvert
spec:
  containers:
  - name: podvert
    image: gcr.io/google_containers/pause-amd64:3.0
    ports:
    - containerPort: 8080
      protocol: TCP
    imagePullPolicy: IfNotPresent
    securityContext:
      privileged: false

```

aakarshg commented:
> Hit another failure
> 
> ```
> jtaleric at bolt in ~/code/snafu on (no branch)▲ $ python2.7 run_snafu.py -t cl -s 1 -p /home/jtaleric/go/src/github.com/openshift/origin/_output/local/bin/linux/amd64/openshift-tests 
> 2019-11-12T15:22:18Z - INFO     - MainProcess - run_snafu: logging level is INFO
> 2019-11-12T15:22:18Z - INFO     - MainProcess - wrapper_factory: identified cl as the benchmark wrapper
> 2019-11-12T15:22:18Z - INFO     - MainProcess - trigger_cluster_loader: running:/home/jtaleric/go/src/github.com/openshift/origin/_output/local/bin/linux/amd64/openshift-tests run-test "[Feature:Performance][Serial][Slow] Load cluster should load the cluster [Suite:openshift]"
> 2019-11-12T15:22:18Z - INFO     - MainProcess - trigger_cluster_loader: from current directory /home/jtaleric/code/snafu
> fail [github.com/openshift/origin/test/extended/cluster/utils.go:192]: Nov 12 15:39:38.437: Error in pod wait... timed out waiting for the condition

@jtaleric looks like the error explains that the issue is with timeout ? maybe increase timeout and retry?

> Can you share the configuration you've used ? as well as the file /home/jtaleric/code/1/cl_output.txt
Also can you please share the output from your previous error ?

aakarshg commented:
> Can you provide a detailed dashboard since you own the CL work? We need someone who understands the workload to build an end to end vs just a tool to warehouse data.
> 
> This current dashboard yields a nested document, do you happen to have a Grafana dashboard using this?

I met with Sebastian Jug ( the author/maintainer of cluster loader ) and captured conversation on an internal [jira card](https://jira.coreos.com/browse/PERFSCALE-459?focusedCommentId=144882&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-144882 )

To summarize about visualization, the only metric we are concerned about is the time taken and we associate it with the test ran ( so ignore the templates just the high level config file ) which we already index with data.

jtaleric commented:
Finally, 

```
jtaleric at bolt in ~/code/snafu on (no branch)▲ $ python2.7 run_snafu.py -t cl -s 1 -p /home/jtaleric/go/src/github.com/openshift/origin/_output/local/bin/linux/amd64/openshift-tests 
2019-11-12T16:08:02Z - INFO     - MainProcess - run_snafu: logging level is INFO
2019-11-12T16:08:02Z - INFO     - MainProcess - wrapper_factory: identified cl as the benchmark wrapper
2019-11-12T16:08:02Z - INFO     - MainProcess - trigger_cluster_loader: running:/home/jtaleric/go/src/github.com/openshift/origin/_output/local/bin/linux/amd64/openshift-tests run-test "[Feature:Performance][Serial][Slow] Load cluster should load the cluster [Suite:openshift]"
2019-11-12T16:08:02Z - INFO     - MainProcess - trigger_cluster_loader: from current directory /home/jtaleric/code/snafu
2019-11-12T16:15:09Z - INFO     - MainProcess - trigger_cluster_loader: completed sample 1 , results in /home/jtaleric/code/1/cl_output.txt
2019-11-12T16:15:09Z - INFO     - MainProcess - trigger_cluster_loader: cl output is {"marker":"cluster_loader_marker","name":"cluster-loader-test","type":"metrics.TestDuration","startTime":"2019-11-12T16:08:06.246060973-05:00","testDuration":423326781857,"Steps":[{"type":"metrics.PodStepDuration","startTime":"2019-11-12T16:08:06.855119589-05:00","totalTime":421121181710,"waitPodsDurations":[13146089393,11221997208,10284781108,13219261350,10088342228,12090228084,11092890120,10092678847,11164832090,10166498705],"rateDelay":0,"rateDelayCount":0,"stepPause":30,"stepPauseCount":10,"syncTime":577}]}
2019-11-12T16:15:09Z - INFO     - MainProcess - trigger_cluster_loader: config file location is /home/jtaleric/code/snafu/podvert.yaml
2019-11-12T16:15:09Z - INFO     - MainProcess - trigger_cluster_loader: configuration is {'ClusterLoader': {'cleanup': True, 'projects': [{'basename': 'podvert', 'num': 1, 'pods': [{'basename': 'podvert', 'num': 100, 'file': 'podvert-pod.yaml'}], 'tuning': 'default', 'ifexists': 'delete'}], 'tuningsets': [{'pods': {'ratelimit': {'delay': 0}, 'stepping': {'stepsize': 10, 'pause': 30, 'timeout': 600}}, 'name': 'default'}]}, 'provider': 'local'}
2019-11-12T16:15:09Z - INFO     - MainProcess - run_snafu: Duration of execution - 0:07:07

```

aakarshg commented:
> Finally,
> 
> ```
> jtaleric at bolt in ~/code/snafu on (no branch)▲ $ python2.7 run_snafu.py -t cl -s 1 -p /home/jtaleric/go/src/github.com/openshift/origin/_output/local/bin/linux/amd64/openshift-tests 
> 2019-11-12T16:08:02Z - INFO     - MainProcess - run_snafu: logging level is INFO
> 2019-11-12T16:08:02Z - INFO     - MainProcess - wrapper_factory: identified cl as the benchmark wrapper
> 2019-11-12T16:08:02Z - INFO     - MainProcess - trigger_cluster_loader: running:/home/jtaleric/go/src/github.com/openshift/origin/_output/local/bin/linux/amd64/openshift-tests run-test "[Feature:Performance][Serial][Slow] Load cluster should load the cluster [Suite:openshift]"
> 2019-11-12T16:08:02Z - INFO     - MainProcess - trigger_cluster_loader: from current directory /home/jtaleric/code/snafu
> 2019-11-12T16:15:09Z - INFO     - MainProcess - trigger_cluster_loader: completed sample 1 , results in /home/jtaleric/code/1/cl_output.txt
> 2019-11-12T16:15:09Z - INFO     - MainProcess - trigger_cluster_loader: cl output is {"marker":"cluster_loader_marker","name":"cluster-loader-test","type":"metrics.TestDuration","startTime":"2019-11-12T16:08:06.246060973-05:00","testDuration":423326781857,"Steps":[{"type":"metrics.PodStepDuration","startTime":"2019-11-12T16:08:06.855119589-05:00","totalTime":421121181710,"waitPodsDurations":[13146089393,11221997208,10284781108,13219261350,10088342228,12090228084,11092890120,10092678847,11164832090,10166498705],"rateDelay":0,"rateDelayCount":0,"stepPause":30,"stepPauseCount":10,"syncTime":577}]}
> 2019-11-12T16:15:09Z - INFO     - MainProcess - trigger_cluster_loader: config file location is /home/jtaleric/code/snafu/podvert.yaml
> 2019-11-12T16:15:09Z - INFO     - MainProcess - trigger_cluster_loader: configuration is {'ClusterLoader': {'cleanup': True, 'projects': [{'basename': 'podvert', 'num': 1, 'pods': [{'basename': 'podvert', 'num': 100, 'file': 'podvert-pod.yaml'}], 'tuning': 'default', 'ifexists': 'delete'}], 'tuningsets': [{'pods': {'ratelimit': {'delay': 0}, 'stepping': {'stepsize': 10, 'pause': 30, 'timeout': 600}}, 'name': 'default'}]}, 'provider': 'local'}
> 2019-11-12T16:15:09Z - INFO     - MainProcess - run_snafu: Duration of execution - 0:07:07
> ```

Cool, so what caused the error the last time ?

jtaleric commented:
> > Finally,
> > ```
> > jtaleric at bolt in ~/code/snafu on (no branch)▲ $ python2.7 run_snafu.py -t cl -s 1 -p /home/jtaleric/go/src/github.com/openshift/origin/_output/local/bin/linux/amd64/openshift-tests 
> > 2019-11-12T16:08:02Z - INFO     - MainProcess - run_snafu: logging level is INFO
> > 2019-11-12T16:08:02Z - INFO     - MainProcess - wrapper_factory: identified cl as the benchmark wrapper
> > 2019-11-12T16:08:02Z - INFO     - MainProcess - trigger_cluster_loader: running:/home/jtaleric/go/src/github.com/openshift/origin/_output/local/bin/linux/amd64/openshift-tests run-test "[Feature:Performance][Serial][Slow] Load cluster should load the cluster [Suite:openshift]"
> > 2019-11-12T16:08:02Z - INFO     - MainProcess - trigger_cluster_loader: from current directory /home/jtaleric/code/snafu
> > 2019-11-12T16:15:09Z - INFO     - MainProcess - trigger_cluster_loader: completed sample 1 , results in /home/jtaleric/code/1/cl_output.txt
> > 2019-11-12T16:15:09Z - INFO     - MainProcess - trigger_cluster_loader: cl output is {"marker":"cluster_loader_marker","name":"cluster-loader-test","type":"metrics.TestDuration","startTime":"2019-11-12T16:08:06.246060973-05:00","testDuration":423326781857,"Steps":[{"type":"metrics.PodStepDuration","startTime":"2019-11-12T16:08:06.855119589-05:00","totalTime":421121181710,"waitPodsDurations":[13146089393,11221997208,10284781108,13219261350,10088342228,12090228084,11092890120,10092678847,11164832090,10166498705],"rateDelay":0,"rateDelayCount":0,"stepPause":30,"stepPauseCount":10,"syncTime":577}]}
> > 2019-11-12T16:15:09Z - INFO     - MainProcess - trigger_cluster_loader: config file location is /home/jtaleric/code/snafu/podvert.yaml
> > 2019-11-12T16:15:09Z - INFO     - MainProcess - trigger_cluster_loader: configuration is {'ClusterLoader': {'cleanup': True, 'projects': [{'basename': 'podvert', 'num': 1, 'pods': [{'basename': 'podvert', 'num': 100, 'file': 'podvert-pod.yaml'}], 'tuning': 'default', 'ifexists': 'delete'}], 'tuningsets': [{'pods': {'ratelimit': {'delay': 0}, 'stepping': {'stepsize': 10, 'pause': 30, 'timeout': 600}}, 'name': 'default'}]}, 'provider': 'local'}
> > 2019-11-12T16:15:09Z - INFO     - MainProcess - run_snafu: Duration of execution - 0:07:07
> > ```
> 
> Cool, so what caused the error the last time ?

It was due to a timeout, which seems to be caused if any pods are still in any state but "Running". 

The problem with this is that CL doesn't seem to cleanup if it hits this state. Not sure if that is expected. However, it would be nice to make this an option (clean:true or clean:false).

aakarshg commented:
> It was due to a timeout, which seems to be caused if any pods are still in any state but "Running".
> 
> The problem with this is that CL doesn't seem to cleanup if it hits this state. Not sure if that is expected. However, it would be nice to make this an option (clean:true or clean:false).

I'd expect that cleanup should happen irrespective when passing cleanup: True, i'll try to reproduce this and if i hit the same error i.e. cleanup not happening open a bug for the same.

jtaleric commented on cluster_loader/cluster_loader.py:
```diff
@@ -0,0 +1,65 @@
+            default='/root/go/src/github.com/openshift/origin/_output/local/bin/linux/amd64/openshift-tests')
+        self.args = parser.parse_args()
+
+        self.server = ""
```
What will this be used for? I don't see it used anywhere?

aakarshg replied:
Me either, will remove this, i used fs-drift/smallfile as the template to add this, so it's a potential leftover :)

aakarshg commented:
@dry923  and @chaitanyaenr please review this when you get a chance and if possible run it to see if it matches the expectations for your output

dry923 commented on cluster_loader/cluster_loader.py:
```diff
@@ -0,0 +1,69 @@
+
+logger = logging.getLogger("snafu")
+
+class cluster_loader_wrapper():
```
Kind of a nit but to keep in line with the other wrappers should the file be cluster_loader_wrapper.py instead of cluster_loader.py?

aakarshg replied:
Nice catch, i'll be going through and fixing all the other wrappers but this is how i think the file naming should be, wrapper is a class not a file if it makes sense.

dry923 commented on cluster_loader/README.md:
```diff
@@ -0,0 +1,42 @@
+# Cluster Loader
+
+To run [cluster loader](https://github.com/openshift/origin/blob/master/test/extended/cluster/) with snafu following environment variables are necessary:
```
"...with snafu _the_ following..."

dry923 commented on cluster_loader/README.md:
```diff
@@ -0,0 +1,42 @@
+# Cluster Loader
```
Not to be overly picky but the below readme feels very disjointed. There are 5 "notes" which discuss additional variables and package requirements. IMHO I would rewrite it to include everything above line 31 as a "Requirements" section (cleaning it up to have a "Package requirements", "Variables", etc, etc) and then have your invoking be the python run_snafu ... down.

aakarshg replied:
done

dry923 commented:
@aakarshg Additional request, however not required atm. Given that I am working through the CI for snafu can you please give some thought to how best to test this. The format I'm working with invokes a ci_test.sh script that would live in the cluster_loader directory and return 0 for success.

chaitanyaenr commented:
Think it would be useful to stream the cluster loader output to know what's it's doing during the run instead of waiting till the benchmark run completes. Thoughts @aakarshg?

chaitanyaenr commented on cluster_loader/README.md:
```diff
@@ -0,0 +1,42 @@
+## Invoking cluster loader through snafu:
+
+Note: Please build the binary openshift-tests by following
+[docs for installing prereqs](https://github.com/openshift/origin/blob/release-4.3/CONTRIBUTING.adoc#installing-prerequisites)  and
```
Think the link is broken.

aakarshg commented:
> Think it would be useful to stream the cluster loader output to know what's it's doing during the run instead of waiting till the benchmark run completes. Thoughts @aakarshg?

So I added notes for this in README.md the output is being streamed to a file so to watch what's gpoing on, you can do a watch on tail on the file, just printing out to screen will make the logs difficult to read.

dry923 commented on cluster_loader/README.md:
```diff
@@ -0,0 +1,46 @@
+
+*KUBECONFIG* and point to kubernetes cluster you want to run cluster loader against
+
+Note: oc binary is also required as cluster loader uses openshift client
```
Add link to get oc binary please :-)

aakarshg replied:
beyond the scope of this PR :)

dry923 commented on cluster_loader/README.md:
```diff
@@ -0,0 +1,46 @@
+
+Step1: Please build the binary openshift-tests by following [building binary](https://github.com/openshift/origin/blob/master/HACKING.md#end-to-end-e2e-and-extended-tests)
+
+Step2: you'll also need to install following python packages as follows:
```
Step 2: You'll also need to install following python packages:

dry923 commented on cluster_loader/README.md:
```diff
@@ -0,0 +1,46 @@
+
+## Requirements
+
+Step1: Please build the binary openshift-tests by following [building binary](https://github.com/openshift/origin/blob/master/HACKING.md#end-to-end-e2e-and-extended-tests)
```
Step 1

chaitanyaenr commented:
> > Think it would be useful to stream the cluster loader output to know what's it's doing during the run instead of waiting till the benchmark run completes. Thoughts @aakarshg?
> 
> So I added notes for this in README.md the output is being streamed to a file so to watch what's gpoing on, you can do a watch on tail on the file, just printing out to screen will make the logs difficult to read.

Think streaming the logs automatically will make it easy for using the wrapper in Automation/Jenkins and even when running it outside Jenkins instead of running tail on the log file on the side during the clusterloader run. Thoughts?

aakarshg commented:
> > > Think it would be useful to stream the cluster loader output to know what's it's doing during the run instead of waiting till the benchmark run completes. Thoughts @aakarshg?
> > 
> > 
> > So I added notes for this in README.md the output is being streamed to a file so to watch what's gpoing on, you can do a watch on tail on the file, just printing out to screen will make the logs difficult to read.
> 
> Think streaming the logs automatically will make it easy for using the wrapper in Automation/Jenkins and even when running it outside Jenkins instead of running tail on the log file on the side during the clusterloader run. Thoughts?

Updated the code to allow for this, please check and try ocne again. I don't have a cluster up and running will launch one and try again. 

Thanks

aakarshg commented:
> > > > Think it would be useful to stream the cluster loader output to know what's it's doing during the run instead of waiting till the benchmark run completes. Thoughts @aakarshg?
> > > 
> > > 
> > > So I added notes for this in README.md the output is being streamed to a file so to watch what's gpoing on, you can do a watch on tail on the file, just printing out to screen will make the logs difficult to read.
> > 
> > 
> > Think streaming the logs automatically will make it easy for using the wrapper in Automation/Jenkins and even when running it outside Jenkins instead of running tail on the log file on the side during the clusterloader run. Thoughts?
> 
> Updated the code to allow for this, please check and try ocne again. I don't have a cluster up and running will launch one and try again.
> 
> Thanks

@chaitanyaenr I've tested it and works fine, additional env variable is needed for Azure which is beyond scope of snafu but just added docs about it
